File: Programmers/[완]모음 사전.js
function make_dict(char, char_arr) {
    if (char.length == 5) {
        return [""];
    }

    var result = []
    for (let element of char_arr) {
        var temp = char + element;
        var child_char = make_dict(temp, char_arr);
        if (child_char.length == 1) {
            result.push(temp);
        }
        else {
            child_char.unshift(temp)
            result.push(...child_char);
        }
    }
    return result;
}

function solution(word) {
    var start_char = word[0];
    var char_arr = ['A', 'E', 'I', 'O', 'U'];
    var dictionary = [];
    var temp;
    for (let char of char_arr) {
        if (char.charCodeAt(0) > start_char.charCodeAt(0)) {
            break;
        }
        temp = make_dict(char, char_arr)
        temp.unshift(char)
        dictionary.push(...temp);
    }
    console.log(dictionary)
    var answer = dictionary.indexOf(word) + 1;
    return answer;
}

var examples = [
    "AAAAE",
    "AAAE",
    "I",
    "EIO"
]

for (example of examples){
    console.log(solution(example));
}

File: Programmers/[완] 모음 사전.py
def make_dict(char, char_list):
    if len(char) == 5:
        return [""]
    result = []
    for element in char_list:
        temp = char + element
        child_char = make_dict(temp, char_list)
        if len(child_char) == 1:
            result += [temp]
        else:
            result += [temp] + child_char
    return result

def solution(word):
    '''
    AA = A + A,
    AAAAE = AAAA + E,
    AAAA = AAA + A
    AAA = AA + A
    AA = A
    '''
    start_char = word[0]
    char_list = ['A', 'E', 'I', 'O', 'U']
    dictionary = []
    for char in char_list:
        if ord(char) > ord(start_char):
            break
        dictionary += [char] + make_dict(char, char_list)
    answer = dictionary.index(word) + 1
    return answer

examples = [
    "AAAAE",
    "AAAE",
    "I",
    "EIO"
]

for example in examples:
    print(solution(example))